### PR TITLE
320 rounds: transaction tail tracker implementation

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary

Move the transaction tail storage to be on a tracker database table txtail

## Test Plan

Write unit tests to verify correctness.

### CODE REVIEW

File: ledger/accountdb.go [hunk 1]
```diff
@@ -1780,6 +1811,49 @@ func performResourceTableMigration(ctx context.Context, tx *sql.Tx, log func(pro
...
+	err = blockDb.Atomic(func(ctx context.Context, blockTx *sql.Tx) error {
+		latest, err := blockLatest(blockTx)
+		if err != nil {
+			return fmt.Errorf("latest block number cannot be retrieved : %w", err)
+		}
+		latestHdr, err := blockGetHdr(blockTx, latest)
+		if err != nil {
+			return fmt.Errorf("latest block header %d cannot be retrieved : %w", latest, err)
+		}
+		maxTxnLife := basics.Round(config.Consensus[latestHdr.CurrentProtocol].MaxTxnLife - 1)
```
Comment (algorandskiy) [inline ledger/accountdb.go:1829]: why -1? Plus, in general we do not have 1001 rounds, only 1000.

**Review by algorandskiy — COMMENTED** [review]
I made a quick review and the following question needed to be addressed:
On migration blockdb has 1000 rounds but we need 1001. This is not a problem if we are few rounds before the protocol upgrade and can re-build the 1001 blocks long tail. But what if one upgrades the node after the protocol upgrade? Their blockdb has only 1000 blocks but the next block to evaluate might require data from 1001 back. This looks OK since most of the stake is upgraded well ahead of the protocol switch and their DB is a right state. But want to hear other opinions.

File: ledger/txtail.go [hunk 2]
```diff
@@ -45,55 +66,57 @@ type txTail struct {
...
+		for i := 0; i < len(txTailRound.Leases); i++ {
+			if consensusParams.SupportTransactionLeases && (txTailRound.Leases[i].Lease != [32]byte{}) {
+				t.recent[old].txleases[ledgercore.Txlease{Sender: txTailRound.Leases[i].Sender, Lease: txTailRound.Leases[i].Lease}] = txTailRound.LastValid[txTailRound.Leases[i].TxnIdx]
 			}
-			if tx.Txn.LastValid > t.lowWaterMark {
-				list := roundsLastValids[tx.Txn.LastValid]
+		}
+
+		for i := 0; i < len(txTailRound.LastValid); i++ {
+			if txTailRound.LastValid[i] > t.lowWaterMark {
```
Comment (algonautshant) [inline ledger/txtail.go:118]: Why not include if less than `t.lowWaterMark`?

Comment (algorandskiy) [reply ledger/txtail.go:118]: this basically repeats the old logic. If last valid is in past the txn is expired and we are not interested in tracking it.

File: ledger/txtail.go [hunk 3]
```diff
@@ -223,3 +309,20 @@ func (t *txTail) putLV(lastValid basics.Round, id transactions.Txid) {
...
 	t.lastValid[lastValid][id] = struct{}{}
 }
+
+func (t *txTail) recentTailHash(offset uint64, maxTxnLife uint64) (crypto.Digest, error) {
+	// preapare a buffer to hash.
+	buffer := make([]byte, (maxTxnLife+1)*crypto.DigestSize)
+	bufIdx := 0
+	t.tailMu.RLock()
+	lastOffset := offset + maxTxnLife
+	if lastOffset > uint64(len(t.roundTailHashes)) {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/txtail.go:319]: Not sure I understand the question but this situation can happen on the a startup of a new network when we do not have all maxTxnLife at all but still want to store some data.

File: ledger/txtail_test.go [hunk 4]
```diff
@@ -152,8 +192,9 @@ func TestTxTailLoadFromDisk(t *testing.T) {
 	partitiontest.PartitionTest(t)
 	var ledger txTailTestLedger
 	txtail := txTail{}
+	require.NoError(t, ledger.Initialize(t))
```
Comment (algorandskiy) [inline ledger/txtail_test.go:195]: must be internal

**Comment by algorandskiy** [comment]
We'll address comments and extend testing in a separate PR

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**